File: Python/binarytree/binarytree.py
import helperfile as hf


# import turtle
class Node:
    def __init__(self, value):
        self.data = value
        self.left = None
        self.right = None


class BinaryTree:
    def __init__(self, value):
        self.root = Node(value)

    def add_recursive(self, current, value):
        if current is None:
            return Node(value)

        if value < current.data:
            current.left = self.add_recursive(current.left, value)
        elif value > current.data:
            current.right = self.add_recursive(current.right, value)

        return current

    def printBoundaryLeft(self, root):
        if root:
            if root.left:
                # to ensure top down order, print the node
                # before calling itself for left subtree
                print(root.data, end=" ")
                self.printBoundaryLeft(root.left)

            elif root.right:
                print(root.data, end=" ")
                self.printBoundaryLeft(root.right)

    def post_order_print(self, current):
        if current is None:
            return
        self.post_order_print(current.left)
        self.post_order_print(current.right)
        print(current.data, end=" ")

    def get_max(self, current):
        if current is None:
            return float("-inf")
        max_value = current.data
        left_max = self.get_max(current.left)
        right_max = self.get_max(current.right)

        return max(max_value, left_max, right_max)

    def get_height(self, current):
        if current is None:
            return 0
        return max(self.get_height(current.left), self.get_height(current.right)) + 1

    def count_all_nodes(self, current):
        if current is None:
            return 0
        return (
            self.count_all_nodes(current.left) + self.count_all_nodes(current.right) + 1
        )

    def count_leaf_nodes(self, current):
        if current is None:
            return 0
        if current.left is None and current.right is None:
            return 1
        return self.count_leaf_nodes(current.left) + self.count_leaf_nodes(
            current.right
        )

    def search_tree(self, current, value):
        if current is None:
            return False
        if current.data == value:
            return True
        return self.search_tree(current.left, value) or self.search_tree(
            current.right, value
        )

    def is_perfect(self, current):
        num = self.count_all_nodes(current)
        height = self.get_height(current) - 1
        return num == 2 ** (height + 1) - 1

    def pre_order_print(self, current):
        if current is None:
            return
        print(current.data, end=" ")
        self.pre_order_print(current.left)
        self.pre_order_print(current.right)

    def in_order_print(self, current):
        if current is None:
            return
        self.in_order_print(current.left)
        print(current.data, end=" ")
        self.in_order_print(current.right)

    def print_inorder_iterative(self, cur):
        nodes = []
        nodes.append((cur, False))

        while nodes:
            current, is_done = nodes.pop()

            if is_done:
                print(current.data, end=" ")
            else:
                if current.right:
                    nodes.append((current.right, False))

                nodes.append((current, True))

                if current.left:
                    nodes.append((current.left, False))
        print("\n")


#

# def draw_tree(turtle, node, x, y, x_dist, y_dist):
#     if node is not None:
#         turtle.penup()
#         turtle.goto(x, y)
#         turtle.pendown()
#         turtle.circle(20)
#         turtle.penup()
#         turtle.goto(x, y - 20)
#         turtle.pendown()
#         turtle.write(str(node.data), align="center", font=("Arial", 12, "normal"))

#         if node.left:
#             turtle.penup()
#             turtle.goto(x - x_dist, y - y_dist)
#             turtle.pendown()
#             turtle.goto(x, y)
#             draw_tree(turtle, node.left, x - x_dist, y - y_dist, x_dist / 2, y_dist - 50)

#         if node.right:
#             turtle.penup()
#             turtle.goto(x + x_dist, y - y_dist)
#             turtle.pendown()
#             turtle.goto(x, y)
#             draw_tree(turtle, node.right, x + x_dist, y - y_dist, x_dist / 2, y_dist - 50)

# Example usage
# if __name__ == "__main__":
# binary_tree = BinaryTree(10)
# binary_tree.add_recursive(binary_tree.root, 5)
# binary_tree.add_recursive(binary_tree.root, 15)
# binary_tree.add_recursive(binary_tree.root, 3)
# binary_tree.add_recursive(binary_tree.root, 7)
# binary_tree.add_recursive(binary_tree.root, 12)
# binary_tree.add_recursive(binary_tree.root, 18)
# screen = turtle.Screen()
# screen.setup(width=800, height=600)
# screen.tracer(0)
# turtle.speed(0)
# turtle.hideturtle()
# draw_tree(turtle, binary_tree.root, 0, 200, 300, 100)
# screen.update()
# turtle.done()


# #
#  def left_boundary(self, cur):
#         if self == None:
#             return
#         if cur.left:
#             print(cur.data, end=" ")

#         self.left_boundary(cur.left)
#         self.left_boundary(cur.right)
# #

# def left_boundary(self, cur):
#         if cur == None:
#             return
#         if cur.left:
#             print(cur.data, end=" ")
#             self.left_boundary(cur.left)

#         else:
#             self.left_boundary(cur.right)

File: Data_Structure/codeblocks/binarytree.cpp
#include <iostream>
#include <cmath>
#include <climits>
#include <algorithm>
#include <queue>
#include "binarytree.h"
using namespace std;



    Node::Node(int val)
    {
        data = val;
        left = nullptr;
        right = nullptr;
    };




    BinaryTree::BinaryTree(int val)
    {
        root = new Node(val);
    }

    bool BinaryTree::isEmpty()
    {
        return root == nullptr;
    }

    Node *BinaryTree::addRecursive(Node *current, int value)
    {
        if (current == nullptr)
        {
            return new Node(value);
        }

        if (value < current->data)
        {
            current->left = addRecursive(current->left, value);
        }
        else if (value > current->data)
        {
            current->right = addRecursive(current->right, value);
        }

        return current;
    }

    void BinaryTree::postOrderPrint(Node *current)
    {
        if (current == nullptr)
        {
            return;
        }
        postOrderPrint(current->left);
        postOrderPrint(current->right);
        cout << current->data << " ";
    }

    Node *BinaryTree::getroot()
    {
        return root;
    }

    int getMax(Node *current)
    {
        if (current == nullptr)
        {
            return INT_MIN;
        }

        int max = current->data;
        int leftMax = getMax(current->left);
        int rightMax = getMax(current->right);

        return std::max(max, std::max(leftMax, rightMax));
    }

    // int getHeight(Node *current)
    // {
    //     if (current == nullptr)
    //     {
    //         return 0;
    //     }
    //
    //     return std::max(getHeight(current->left), getHeight(current->right)) + 1;
    // }

    int BinaryTree::getHeight(Node *root)
    {
        if (root == nullptr)
            return 0;
        else
        {
            int leftHeight = getHeight(root->left);
            int rightHeight = getHeight(root->right);

            return max(leftHeight, rightHeight) + 1;
        }
    }

    int BinaryTree::countAllNodes(Node *current)
    {
        if (current == nullptr)
        {
            return 0;
        }

        return countAllNodes(current->left) + countAllNodes(current->right) + 1;
    }

    int BinaryTree::countLeafNodes(Node *current)
    {
        if (current == nullptr)
        {
            return 0;
        }

        if (current->left == nullptr && current->right == nullptr)
        {
            return 1;
        }

        return countLeafNodes(current->left) + countLeafNodes(current->right);
    }

    bool BinaryTree::searchTree(Node *current, int val)
    {
        if (current == nullptr)
        {
            return false;
        }

        if (current->data == val)
        {
            return true;
        }

        return searchTree(current->left, val) || searchTree(current->right, val);
    }

    bool BinaryTree::isPerfect(Node *current)
    {
        int num = countAllNodes(current);
        int height = getHeight(current) - 1;

        return num == std::pow(2, height + 1) - 1;
    }

    void BinaryTree::preOrderPrint(Node *current)
    {
        if (current == nullptr)
        {
            return;
        }
        std::cout << current->data << " ";
        preOrderPrint(current->left);
        preOrderPrint(current->right);
    }

    void BinaryTree::inOrderPrint(Node *current)
    {
        if (current == nullptr)
        {
            return;
        }
        inOrderPrint(current->left);
        std::cout << current->data << " ";
        inOrderPrint(current->right);
    }

    int BinaryTree::Get_Diametar(Node *current)
    {

        int diametar = 0;
        Get_Hight_for_diametar(current, diametar);
        return diametar;
    }

    int BinaryTree::Get_Hight_for_diametar(Node *current, int &diametar)
    {

        if (!current)
        {
            return 0;
        }

        int h1 = Get_Hight_for_diametar(current->left, diametar);
        int h2 = Get_Hight_for_diametar(current->right, diametar);

        diametar = max(diametar, (h1 + h2));
        return 1 + max(h1, h2);
    }

    void BinaryTree::Level_by_Level_iterative()
    {
        int level = 0;
        queue<Node *> qu;
        qu.push(getroot());

        while (!qu.empty())
        {

            cout << "Level " << level << ": ";
            int sz = qu.size();
            while (sz--)
            {
                Node *current = qu.front();
                qu.pop();

                cout << current->data << " ";
                if (current->left)
                    qu.push(current->left);
                if (current->right)
                    qu.push(current->right);
            }
            level++;
            cout << endl;
        }
    }

    void BinaryTree::Level_by_Level_recursive(Node *cur)
    {

        int sz = getHeight(cur);
        for (int level = 0; level < sz; level++)
        {
            Printlevel(level, cur);
        }
    }

   void BinaryTree::Printlevel(int level, Node *cur) {
    if (level == 0) {
        cout << cur->data << " ";
    } else if (level > 0) {
        if (cur->left) {
            Printlevel(--level, cur->left);
        }
        if (cur->right) {
            Printlevel(--level, cur->right);
        }
    }
}




















    // void print_level(Node *cur, int level)
      // {

    //     if (level == 0)
    //     {
    //         cout << cur->data << " ";
    //     }
    //     else if (level > 0)
    //     {
    //         if (cur->left)
    //         {
    //             print_level(cur->left, level - 1);
    //         }
    //         if (cur->right)
    //         {
    //             print_level(cur->right, level - 1);
    //         }
    //     }
    // }

    // void Level_by_Level_recursive(Node *current)
    // {
    //     int th = getHeight(current);
    //     for (int i = 0; i < th; i++)
    //     {
    //         print_level(current, i);
    //     }
    //     cout << endl;
    // }

